### DISCUSSION TITLE
[design decision] Optional trailing comma's

### DISCUSSION THREAD

**Discussion by mtijink** [tabular-json/tabular-json#5] [Ideas]
Interesting idea! I see some parallels with [JSON5](https://json5.org/), which includes features I like (optional trailing comma's and hexadecimal numbers). Maybe this can serve as further inspiration?

**Comment by josdejong** [comment]
Thanks for your inputs Matthijs! 

The Tabular-JSON format focuses to be a good data format, whereas JSON5 aims at being a configuration file. 

I want to be careful putting "a lot of handy extra stuff" into Tabular-JSON: that makes the standard more complex to implement and thus less likely to be adopted in various environments. The reason that JSON and CSV are so popular is (amongst others) because they are so dead simple. So we should carefully weigh the added complexity against the added value.

I expect that trailing commas will not add much to the complexity of the standard, it's interesting to see what happens if we add that. Besides convenience and being git-friendly, it can be interesting too because allowing trailing commas makes the standard better streamable.

I have the feeling that hexadecimal numbers is quite an edge case, but if there is concrete interest in it let's discuss in a separate thread, OK?

Complexity: low
Added value: medium

**Reply by josdejong** [reply]
I've read up a bit more to get a clear view on the pros and cons of supporting trailing commas.

Pros:

- Cleaner git diffs
- Easier code manipulations (to add or rearrange items)
- Makes it easier for code generators
- Many programming languages support it, it is quite popular

Cons:

- Added complexity
- Inconsistent usage
- Ambiguity (multiple ways to do the same thing)
- Inconsistency issues across languages (not all languages support it)
- For the human eye, it just looks ugly: it reads like an unfinished sentence, and

Background articles:

- https://medium.com/@nikgraf/why-you-should-enforce-dangling-commas-for-multiline-statements-d034c98e36f8
- https://aykev.dev/blog/trailing-commas.html
- https://devblogs.microsoft.com/oldnewthing/20240209-00/?p=109379
- https://medium.com/@8acking/is-trailing-comma-really-necessary-reasons-why-it-might-be-better-not-to-use-it-f5b2d7556d5e

Conclusion:

Let's add support for trailing comma's to Tabular-JSON. I think the pros outweigh the cons.

**Reply by josdejong** [reply]
Implemented via #11
